### PR TITLE
specdec(recipe): add MiniMax-M2.7-DFlash streaming multi-node pipeline

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
## Summary

- Add `hf_streaming_dflash_multi_node.yaml` for MiniMax-M2.7 (229B MoE) streaming DFlash training
- 2 serve replicas (TP=4, whole node) + 2 trainer nodes (4 GPU each) over NIXL RDMA hidden-state transport
- Capture IDs `[2,17,32,47,62,64]` from `build_target_layer_ids(64, 5)` + final layer output
- MiniMax-specific: trust_remote_code, FSDP2 via accelerate config, mask_token=200054, YaRN rope_scaling factor=48
- Topology matches Kimi-K2.5 large-MoE streaming recipe

Resolves OMNIML-5221

## Test plan

- [ ] Dry-run validation (`uv run launch.py --yaml ... --dry-run`)
- [ ] Server-only smoke on CW-DFW (task_1 with `training.max_steps=1`)
- [ ] Full streaming training run

Signed-off-by: Ye Yu <yeyu@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Added a multi-node launcher configuration for MiniMax-M2.7 streaming training with speculative decoding.
  * Added an end-to-end workflow for conversational data preparation, distributed training, checkpoint export, and vLLM smoke testing.
  * Supports configurable serving and training settings across multiple nodes.

* **Enhancements**
  * Applies Transformers version overrides only to trainer or single-node environments.
  * Preserves the serving environment’s Transformers version on dedicated serving nodes.
  * Uses the launcher’s built-in bfloat16 mixed-precision configuration for training.

### CODE REVIEW

File: tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      - dflash.dflash_export_rope_scaling.type=yarn
+      - dflash.dflash_export_rope_scaling.factor=48.0
+      - dflash.dflash_export_rope_scaling.original_max_position_embeddings=4096
+      - dflash.dflash_export_rope_scaling.beta_fast=1.0
+      - dflash.dflash_export_rope_scaling.beta_slow=1.0
+      - dflash.dflash_export_rope_scaling.mscale=1.0
+      - dflash.dflash_export_rope_scaling.mscale_all_dim=1.0
+    environment:
+      - HF_MODEL_CKPT: <<global_vars.hf_model>>
+      - EAGLE_CAPTURE_IDS: "[2,17,32,47,62,64]"
```
Comment (h-guo18) [inline tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml]: QQ: minimax seems to have 62 layers. What does 64 means here?

Comment (yeyu-nvidia) [reply tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml]: Good catch — MiniMax-M2.7 has 62 hidden layers, not 64. I miscounted from the prior DFlash M3 work (which is 64 layers). Fixed in b71297f: recalculated `EAGLE_CAPTURE_IDS` from `build_target_layer_ids(62, 5)` → `[2,17,31,45,60,62]`.

**Comment by yeyu-nvidia** [comment]
@h-guo18 Thanks for catching the layer count — fixed in b71297f (62 not 64, capture IDs updated to `[2,17,31,45,60,62]`).

Two additional commits since your review:
- **8ea2ced**: `OVERRIDE_TRANSFORMERS` support in `train_eagle_streaming.sh` (was missing, only `dflash_online_training.sh` had it)
- **915afca**: Role-aware dependency install — serve nodes now skip `requirements.txt` and transformers override entirely (vLLM v0.24+ hard-rejects transformers v4 at import)

Smoke-tested on CW-DFW (cicd_1782759203): training completed 1 step successfully with loss=12.76. All CI checks pass. Ready for re-review when you get a chance.

**Comment by yeyu-nvidia** [comment]
**What it is:** a MiniMax-M2.7-DFlash streaming multi-node launcher recipe — two files, a YAML example plus 8 lines in `train_eagle_streaming.sh`. No Python.

**On the red `codecov/project`:** it is not a real coverage regression and it will not clear on a re-run. Because the PR changes no Python, the example lanes are correctly path-gated off and upload no coverage — `examples-llm_qat`, `llm_sparsity`, `megatron_bridge`, `specdec_bench`, `speculative_decoding`, `torch_onnx`, `torch_trt` all report flag `?`, and `unit`/`regression` show `<ø>` (no lines in scope changed). Project coverage is therefore computed from a partial flag set against a base that had all of them.

It is also not a required check — two of the last 25 merged PRs in this repo went in with `codecov/project` red — so this is mergeable on approval alone. Looks like a repo-level codecov config gap for docs/config-only PRs rather than anything this PR can fix.

Branch is current with main, everything else green.

**Comment by yeyu-nvidia** [comment]
@kevalmorabia97 — redirecting this one to you for review.

**What it is:** a MiniMax-M2.7-DFlash streaming multi-node launcher recipe — two files, a YAML example plus 8 lines in `train_eagle_streaming.sh`. No Python.

**On the red `codecov/project`:** not a real coverage regression, and it will not clear on a re-run. The PR changes no Python, so the example lanes are correctly path-gated off and upload no coverage — `examples-llm_qat`, `llm_sparsity`, `megatron_bridge`, `specdec_bench`, `speculative_decoding`, `torch_onnx`, `torch_trt` all report flag `?`, while `unit`/`regression` show `<ø>` (no lines in scope changed). Project coverage is therefore computed from a partial flag set against a base that had all of them.

It is also not a required check — two of the last 25 merged PRs here went in with `codecov/project` red — so this is mergeable on approval alone. Looks like a repo-level codecov config gap for docs/config-only PRs rather than anything this PR can fix.

Branch is current with main; everything else green.

**Review by ChenhanYu — APPROVED** [review]

File: tools/launcher/common/eagle3/train_eagle_streaming.sh [hunk 2]
```diff
@@ -95,6 +95,14 @@ fi
...
 pip install --no-cache-dir -r modules/Model-Optimizer/examples/speculative_decoding/requirements.txt
 pip install --no-cache-dir 'datasets' 'huggingface-hub>=1.2.1'
+
+# Some trust_remote_code models pin an older transformers (e.g. MiniMax-M2.7
+# needs 4.57.x whose modeling code is incompatible with the 5.x that the
+# requirements pull in). Must run AFTER the requirements install to win.
+if [ -n "${OVERRIDE_TRANSFORMERS:-}" ]; then
+    pip install --no-cache-dir "transformers==${OVERRIDE_TRANSFORMERS}"
+fi
+
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply tools/launcher/common/eagle3/train_eagle_streaming.sh:122]: Confirmed and fixed in `73bffbc1`. This was a real one, and introduced by this PR.

Verified the structure: the 8 lines added here sit at ~line 98 in the provisioning block, while `NODEID`/`NNODES` were not derived until ~line 262 and the serve/trainer dispatch is at ~line 295. So every node ran the downgrade, and on this YAML's own topology (`nodes: 4`, `SERVE_NODES: 2`, container `vllm/vllm-openai:nightly`) nodes 0-1 would have been downgraded to 4.57.1 and then started `vllm serve`. You are also right that the trainer smoke test could not have caught it — the trainer is precisely the role that wants the pin.

On the stale-note question: the role-aware gating from that earlier comment is genuinely absent at this head, so it was lost rather than merely undocumented.

Took your suggested shape, including hoisting the derivation so both sites share it rather than duplicating. One deliberate difference: the gate is `NNODES > 1 && NODEID < SERVE_NODES`, so single-node runs are untouched — there one environment hosts both `vllm serve` and the trainer, so there is no role to gate on and the trainer still needs the pin. Verified across the topologies:

| NNODES | NODEID | SERVE_NODES | role | result |
|---|---|---|---|---|
| 4 | 0,1 | 2 | serve | skipped (logged) |
| 4 | 2,3 | 2 | trainer | downgraded |
| 1 | 0 | 1 | co-located | downgraded |

Left alone deliberately: the unconditional `pip install -r requirements.txt` above also pulls transformers 5.x onto serve nodes. That predates this PR, so changing it here would be a silent behaviour change to something this PR did not touch — worth its own fix if it is biting.

File: tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      - SERVE_MAX_NUM_SEQS: "4"
+      - SERVE_GPU_MEM_UTIL: "0.8"
+      - SERVE_READY_TIMEOUT: "2400"
+      - SERVE_EXTRA_ARGS: "--trust-remote-code"
+      - VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS: "1200"
+      - VLLM_ENGINE_ITERATION_TIMEOUT_S: "1200"
+      - OVERRIDE_TRANSFORMERS: "4.57.1"
+      - ACCELERATE_CONFIG: examples/MiniMax/MiniMax-M2.7-DFlash/accelerate_fsdp2_hybrid.yaml
+      - PATCH_FSDP2_BUFFERS_TF457: "1"
+      - MIXED_PRECISION: "no"
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml]: Verified and fixed in `fc8b4f3a`. Both are gone.

Confirmed the dataflow: `grep -rn "ACCELERATE_CONFIG\|MIXED_PRECISION" tools/launcher/common/eagle3/` returns nothing; `train_eagle_streaming.sh:246` launches `launch_train.sh`, and that script says so itself at line 23 ("mixed_precision is fixed to bf16") before building `accelerate launch --mixed_precision bf16` at line 76 with no `--config_file`. So both variables were dead, and leaving them implied FSDP2 sharding this path never configures.

Removed rather than rewired — wiring them up would be a behaviour change to the streaming launcher, which is beyond what this recipe should carry.

File: tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      - training.per_device_train_batch_size=2
+      - training.learning_rate=1.2e-3
+      - training.warmup_steps=100
+      - training.logging_steps=100
+      - training.save_steps=400
+      - training.ddp_timeout=3600
+      # dp_shard_size=1 stops main.py from creating a ParallelismConfig; the
+      # accelerate config below owns FSDP2 sharding.
+      - training.dp_shard_size=1
+      - training.bf16=false
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml]: Fixed in `fc8b4f3a`, following from the previous thread.

`training.bf16=false` was indeed the online recipe's pairing with `MIXED_PRECISION=no`. With `MIXED_PRECISION` ignored here and accelerate forcing bf16, keeping it gave fp32 draft params under bf16 autocast with `TrainingArguments` disagreeing with `ACCELERATE_MIXED_PRECISION` — so it is dropped and bf16 follows the recipe default.

`dp_shard_size=1` kept, but the comment no longer claims an accelerate config owns sharding; it now says why the value is stated explicitly even though it matches the `dflash.yaml` default.

File: tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+#   - 229B MoE, trust_remote_code, needs TP=4 per serve replica (each replica = 1 node).
+#   - Custom chat_template_train.jinja for the MiniMax chat format.
+#   - The DFlash draft is Qwen3-architecture (5 layers); target/draft architectures
+#     are independent by design.
+#   - Mask token id 200054 (reserved row in MiniMax-M2.7 embedding).
+#   - YaRN rope_scaling at export for long context (factor = 196608/4096 = 48).
+#
+# Capture ids: build_target_layer_ids(num_orig=62, num_draft=5)=[1,16,30,44,59]
+# -> +1 for embedding = [2,17,31,45,60], append final layer 62.
+# 6 captured = 5 aux layers + final output for self-logit distillation.
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml:21]: Added in `fc8b4f3a` — thanks for checking the derivation independently.

Took the Kimi wording almost verbatim, since the case is identical: layer 62 is the true final hidden and needs the aux-capture fix (vllm#46788); without it the capture silently degrades to the 2nd-to-last layer and surfaces only as a worse acceptance length much later. That silent-degradation property is exactly why it belongs in the header rather than nowhere.

File: tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      - ACCELERATE_CONFIG: examples/MiniMax/MiniMax-M2.7-DFlash/accelerate_fsdp2_hybrid.yaml
+      - PATCH_FSDP2_BUFFERS_TF457: "1"
+      - MIXED_PRECISION: "no"
+    slurm_config:
+      _factory_: "slurm_factory"
+      nodes: 4
+      segment: 4
+      ntasks_per_node: 1
+      gpus_per_node: 4
+      container: vllm/vllm-openai:nightly
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply tools/launcher/examples/MiniMax/MiniMax-M2.7-DFlash/hf_streaming_dflash_multi_node.yaml:127]: Agreed in principle, but I have deliberately not changed it, because I would be guessing.

The point stands: `nightly` re-rolls all three properties you list, and the transformers-4-vs-5 rejection is that failure mode. The problem is that the *correct* pin is whichever image this recipe was actually validated against, and that is not recoverable from the tree. The repo currently spans `v0.21.0`, `v0.22.0`, `v0.22.1`, `qwen3_5-cu130`, `latest` and `nightly`, with 11 recipes on `nightly` — and the Kimi pin you cite (`v0.22.0`) is justified by Kimi-specific serve behaviour that says nothing about MiniMax-M2.7 or about which build carries vllm#46788.

Pinning to a version that lacks the aux-capture fix would silently cap acceptance length — the failure the header now warns about — and pinning to one that regresses MiniMax-M2 serving would break the recipe outright. So this needs the digest or tag from the run that actually worked.

@yeyu-nvidia — if you have the image from the validated run, that is the pin. Worth noting it is a repo-wide pattern rather than unique to this recipe, so it may deserve its own cleanup.
